00-bleeding-edge-java/04-parsing.md

```diff
@@ -0,0 +1,222 @@
+[⬆️ Top](00-introduction.md) • [⬅️ Previous](03-printing.md) • [Next ➡️](05-deserialization-01.md)
+
+# Bleeding Edge Java
+
+_Note: this is Part 4 of this series. Please [start at the introduction](00-introduction.md) if you haven't already_.
+
+## Part 4 - Parsing
+
+Parsing is the process of interpreting free form JSON text and converting it into a stream of JSON tokens. The
+[JSON Spec](https://www.rfc-editor.org/rfc/rfc8259) defines the manner in which parsing should be done: what
+constitutes whitespace, escape sequences in strings, etc. In short, take a stream of characters and interpret
+those characters as a stream of JSON tokens. We can define a Java interface for this:
+
+```java
+public interface JsonParser
+{
+    Stream<JsonToken> parse(IntStream stream);  // Note: Java Strings can be streamed via the chars() method
+}
+```
+
+Instead of a normal Java `Stream` we need to use an `IntStream` as we're dealing with primitive values: characters. Java does not have
+a `CharStream` so `IntStream` will suffice.
+
+There are a number of wrinkles we'll have to deal with when interpreting characters:
+
+- We sometimes need to look ahead before we know what token type is implied (for literals, strings, numbers)
+- Our library introduces an object name token. This means that when we parse a string we need to wait for the next token. If it's a `:` we have to convert the string into an object name.
+- Number tokens can possibly end with a parseable character: `,` `[` `{`. When this happens we need to reprocess that character.
+
+Here is the general flow for parsing JSON text:
+
+![](assets/parse-flow.png)
```

Comment:
There are two blank white squares at the top left and bottom right corners of the image whose purpose is unclear.

Reply:
Fixed